server-cloud/server/auth.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return c.String(http.StatusInternalServerError, "Failed to verify ID Token.")
+	}
+
+	var profile map[string]interface{}
+	if err := idToken.Claims(&profile); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["access_token"] = token.AccessToken
+	sess.Values["profile"] = profile
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	// Redirect to logged in page.
+	return c.Redirect(http.StatusTemporaryRedirect, "/auth/user")
+}
+
+func (s *Server) logoutCallback(c echo.Context) error {
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+	}
+
+	sess.Values["access_token"] = nil
+	sess.Values["profile"] = nil
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	return c.String(http.StatusOK, "user logged out")
+}
+
+func (s *Server) logout(c echo.Context) error {
+	logoutUrl, err := url.Parse("https://" + s.conf.Auth0Domain + "/v2/logout")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	scheme := "http"
+	if c.Request().TLS != nil {
+		scheme = "https"
+	}
+
+	returnTo, err := url.Parse(scheme + "://" + c.Request().Host + "/auth/logout/callback")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	parameters := url.Values{}
+	parameters.Add("returnTo", returnTo.String())
+	parameters.Add("client_id", s.conf.Auth0ClientID)
+	logoutUrl.RawQuery = parameters.Encode()
+
+	return c.Redirect(http.StatusTemporaryRedirect, logoutUrl.String())
+}
+
+func (s *Server) user(c echo.Context) error {
+	sess, err := session.Get(authSessionName, c)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	profile := sess.Values["profile"]
+	return c.JSON(http.StatusOK, profile)
+}
+
+func GetAuth(c context.Context, conf Config) (*Authenticator, error) {
+	provider, err := oidc.NewProvider(
+		c,
+		"https://"+conf.Auth0Domain+"/",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	config := oauth2.Config{
+		ClientID:     conf.Auth0ClientID,
+		ClientSecret: conf.Auth0ClientSecret,
+		RedirectURL:  conf.Auth0CallbackURL,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "profile"},
+	}
+
+	return &Authenticator{
+		Provider: provider,
+		Config:   config,
+	}, nil
+}
+
+func generateRandomState() (string, error) {
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+
+	state := base64.StdEncoding.EncodeToString(b)
+	return state, nil
+}
+
+// IsAuthenticated is a middleware that checks if
+// the user has already been authenticated previously.
+func (s *Server) IsAuthenticated(next echo.HandlerFunc) echo.HandlerFunc {
```

Comment:
This function is internal, so probably better to make it lowercase. And it doesn't have to be a member of `(s *Server)` since it doesn't access the server. I would make it `func authenticatedMiddleware(next ...`